src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithUNION.java

```diff
@@ -0,0 +1,80 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.algebra.Op;
+import org.apache.jena.sparql.algebra.op.OpFilter;
+import org.apache.jena.sparql.algebra.op.OpTriple;
+import org.apache.jena.sparql.algebra.op.OpUnion;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.expr.E_Equals;
+import org.apache.jena.sparql.expr.E_LogicalAnd;
+import org.apache.jena.sparql.expr.Expr;
+import org.apache.jena.sparql.expr.ExprVar;
+import org.apache.jena.sparql.expr.nodevalue.NodeValueNode;
+
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.SPARQLRequestImpl;
+import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.query.impl.QueryPatternUtils;
+import se.liu.ida.hefquin.engine.query.impl.SPARQLGraphPatternImpl;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+public class ExecOpBindJoinSPARQLwithUNION extends ExecOpGenericBindJoin<TriplePattern,SPARQLEndpoint> {
+
+	private Set<Var> varsInTP;
+
+	public ExecOpBindJoinSPARQLwithUNION( final TriplePattern query, final SPARQLEndpoint fm) {
+		super(query, fm);
+		varsInTP = QueryPatternUtils.getVariablesInPattern(query);
+	}
+
+	@Override
+	protected Iterable<? extends SolutionMapping> fetchSolutionMappings( final Iterable<SolutionMapping> solMaps,
+			final ExecutionContext execCxt) {
+		final Op op = createUnion(solMaps);
+		final SPARQLGraphPattern pattern = new SPARQLGraphPatternImpl(op);
+		final SPARQLRequest request = new SPARQLRequestImpl(pattern);
+		final SolMapsResponse response = execCxt.getFederationAccessMgr().performRequest(request, fm);
+		return response.getSolutionMappings();
+	}
+
+	protected Op createUnion(final Iterable<SolutionMapping> solMaps) {
+		Op tp = new OpTriple(query.asJenaTriple());
+		if (varsInTP.isEmpty()) return tp;
+		Op union = null;
+		for ( final SolutionMapping s : solMaps) {
+			final Binding b = SolutionMappingUtils.restrict(s.asJenaBinding(), varsInTP);
+			if (b.size() == 0) continue;
+			Expr conjunction = null;
+			final Iterator<Var> vars = b.vars(); 
+			while (vars.hasNext()) {
+				final Var v = vars.next();
+				final Node uri = b.get(v);
+				final Expr expr = new E_Equals(new ExprVar(v), new NodeValueNode(uri));
```

Comment:
There are cases in which this may result in incorrect intermediate results produced by this operator (more precisely, intermediate results that contain spurious duplicates).

The issue is twofold:
1. The call of `restrict` above may give us duplicates. Then, for each such duplicate, you add an identical copy of `conjunction` to the UNION pattern. Now, for every solution mapping that is a solution for this `conjunction`, the SPARQL endpoint returns as many duplicates as there are copies of this `conjunction` in the UNION pattern. Joining these duplicates with the given solution mappings in `solMaps` (which happens in the superclass of this class here), will result incorrect duplicates coming out of this operator. The fix for this issue is to avoid adding multiple identical copies of the same `conjunction` to the UNION pattern.
2. There are cases  in which things are even more complicated. Unfortunately! As an example, consider `varsInTP` contains two variables, say ?v1 and ?v2, and there are two input solution mappings s1 and s2 such that s1 has a value for both ?v1 and ?v2, respectively, whereas s2 has a value for ?v1 but not for ?v2.  Also assume that the value for ?v1 is the same in both solution mappings; i.e., s1(?v1)=s2(?v1). Hence, we may have something like the following: s1(?v1)=s2(?v1)=uri1, s1(?v2)=uri2, and s2(?v2) is undefined. Then, the resulting UNION pattern might be as follows:
```
{
   # this subpattern is from s1
   ?v1 :p ?v2 . FILTER( ?v1=uri1 && ?v2=uri2 )
}
UNION
{
   # this subpattern is from s2
   ?v1 :p ?v2 . FILTER( ?v1=uri1 )
}
```
Note that every triple that every solution for the first subpattern is also a solution for the second subpattern. Hence, we may introduce spurious duplicates when join the result of this query with the two input solution mappings.

Do you have a good idea to avoid these issues? ("good" in the sense of i) easy to implement and ii) not too problematic from a performance perspective)

If not, I suggest we ignore the UNION-based bind join. That is, I will copy the explanation from above into #34, mark #34 as "won't-fix", and we delete this PR.

Reply:
Another option is to implement a restricted version of this operator which will explicitly be only for cases where such issues are guaranteed not to occur (which can be checked based on the `ExpectedVariables` of the input operator, together with the variables of the given triple pattern). Perhaps that's a good compromise!?

Reply:
For the first issue, we can maintain then a set of `conjunction` before creating the UNION Op object.

Given such a set, the second issue might be addressed by simply comparing all pairs of filters, but as you pointed out it would be very costly.

As for checking the `ExpectedVariables` I believe we would incur in a similar problem (having to check every pair of possibilities), unless we check that the variables in the mappings and triple pattern are only certain variables (but in that case, the mappings are not bound to use all of the certain variables, so that might be the check?).

Reply:
The idea with checking the ExpectedVariables is to check that the set of *possible* variables of the input operator is empty. If that's the case, then every solution mapping from the input operator is guaranteed to have the same variables and, thus, the weird special cases are ruled out. Right? (please double check my reasoning here) 

Of course, it would still be necessary to avoid duplicates as per the first of the two points in my comment above.

Reply:
The `ExpectedVariables` do not really "restrict[] the solution mappings fed to a tpAdd operator." Instead, the `ExpectedVariables` (of the operator that is the child of the tpAdd operator) give us information about these solution mappings. That's the main purpose of this data structure. It tells us what is expected to come out of an operator and, thus, it enables us to know what is expected to come in to the parent operator (which is a tpAdd operator in this case).

I was also thinking about the other two types of bind join algorithms for SPARQL endpoints.

For the FILTER-based version, I think there is no actual problem regarding correctness. The duplicates in the FILTER are not nice in terms of unnecessary data shipping from the engine to the SPARQL endpoint but, at least, they do not cause spurious duplicates in the query result (every triple that matches the triple pattern can satisfy the filter condition only once, even if it may satisfy multiple disjuncts of the condition).

For the VALUES-based version, unfortunately, you are right. There, spurious duplicates may occur.

Reply:
I'm seeing that we never call the `createExecOp` function in the tests, thus the check of the expected variables will never occur, thus I will have no information on how the check will affect the test queries. 

Maybe the method `createExecOpForTest` should create a new `PhysicalOpBindJoinWithVALUES` object and ask it to `createExecOp`. Would that be okay?

Reply:
Good observation! Yes, do it as you propose.